soroban-env-host/src/host/metered_vector.rs

```diff
@@ -47,22 +46,37 @@ where
     A: MeteredClone,
 {
     pub fn new(budget: &Budget) -> Result<Self, HostError> {
-        Self::charge_new(0, budget)?;
-        Self::from_vec(Vec::new())
+        Self::from_vec(Vec::new(), budget)
+    }
+
+    // Constructs a new, empty `MeteredVector` with at least the specified capacity.
+    // This is purely used for the cost calibration of allocating host memory.
+    // Do *not* use it for construction, since `MeteredVector` is immutable,
+    // the allocation will be wasted.
+    #[cfg(any(test, feature = "testutils"))]
+    pub fn with_capacity(capacity: usize, budget: &Budget) -> Result<Self, HostError> {
+        super::metered_clone::charge_heap_alloc::<A>(capacity as u64, budget)?;
+        Self::from_vec(Vec::with_capacity(capacity), budget)
     }
 
     pub fn from_array(buf: &[A], budget: &Budget) -> Result<Self, HostError> {
         // we may temporarily go over budget here.
         let vec: Vec<A> = buf.into();
         vec.charge_deep_clone(budget)?;
-        Self::from_vec(buf.into())
+        Self::from_vec(vec, budget)
     }
 
-    pub fn from_vec(vec: Vec<A>) -> Result<Self, HostError> {
-        // No charge here: vector already allocated, charge happened in caller.
+    pub fn from_vec(vec: Vec<A>, budget: &Budget) -> Result<Self, HostError> {
+        // Only charge for the new vector, assuming allocation cost has been covered
+        // by the caller from the outside.
+        Self::charge_new(1, budget)?;
         Ok(Self { vec })
     }
 
+    pub fn from_free_vec(vec: Vec<A>) -> Self {
```

Comment:
This shouldn't be necessary -- the diagnostics path that constructs vectors is in a lambda that has turned off budget metering altogether.

Reply:
Yes you are right. I initially thought we would run into interior mutability trouble with budget but I was wrong. Removed.